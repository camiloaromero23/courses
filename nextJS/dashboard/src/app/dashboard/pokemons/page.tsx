import { PokemonGrid, PokemonsResponse, SimplePokemon } from "@/pokemons";

export const metadata = {
  title: "Pokemons list",
  description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
}

const getPokemons = async (
  limit = 20,
  offset = 0,
): Promise<SimplePokemon[]> => {
  const res = await fetch(
    `https://pokeapi.co/api/v2/pokemon?limit=${limit}&offset=${offset}`,
  );

  const data: PokemonsResponse = await res.json();

  const pokemons = data.results.map((pokemon) => ({
    id: pokemon.url.split("/").at(-2)!,
    name: pokemon.name,
  }));

  // throw new Error("Not implemented");
  // throw notFound();

  return pokemons;
};

export default async function PokemonsPage() {
  const pokemons = await getPokemons(150);

  return (
    <div className="flex flex-col p-2">
      <span className="text-5xl my-2">
        Pokemons list <small className="text-blue-500">static</small>
      </span>

      <PokemonGrid pokemons={pokemons} />
    </div>
  );
}
