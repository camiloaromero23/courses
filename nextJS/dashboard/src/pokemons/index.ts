export { PokemonGrid } from "./components/PokemonGrid";
export { FavoritePokemons } from "./components/FavoritePokemons";

export type { PokemonsResponse } from "./interfaces/pokemonsResponse";
export type { SimplePokemon } from "./interfaces/simplePokemon";
export type { Pokemon } from "./interfaces/pokemon";
