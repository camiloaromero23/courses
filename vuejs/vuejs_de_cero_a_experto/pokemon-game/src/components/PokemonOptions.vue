<template>
  <div class="options-container">
    <ul>
      <li
        v-for="pokemon in pokemons"
        :key="pokemon.id"
        :disabled="optionsDisabled ? disabled : ''"
        @click="hasAnswer(pokemon.id)"
        :class="{
          disabled: !optionsDisabled,
          'correct-pokemon': pokemonId === pokemon.id && !optionsDisabled,
          'wrong-pokemon':
            pokemonId !== selectedPokemon &&
            pokemon.id === selectedPokemon &&
            !optionsDisabled,
        }"
      >
        {{ pokemon.name }}
      </li>
    </ul>
  </div>
</template>

<script>
export default {
  name: 'PokemonPicture',
  props: {
    pokemons: {
      type: Array,
      required: true,
    },
    optionsDisabled: {
      type: Boolean,
      required: true,
    },
    pokemonId: {
      type: Number,
    },
    selectedPokemon: {
      type: Number,
    },
  },
  methods: {
    hasAnswer(pokemonId) {
      if (this.optionsDisabled) {
        this.$emit('pokemonSelected', pokemonId);
      }
    },
  },
};
</script>

<style scoped>
ul {
  list-style-type: none;
}
li {
  background-color: White;
  border-radius: 5px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  cursor: pointer;
  margin-bottom: 10px;
  width: 250px;
}

li:hover {
  background-color: rgba(0, 0, 0, 0.05);
}

.disabled:hover {
  cursor: not-allowed;
  background-color: White;
}

.correct-pokemon {
  border: 0.15rem solid Green;
}

.wrong-pokemon {
  border: 0.15rem solid Red;
}

.options-container {
  display: flex;
  justify-content: center;
}
</style>
