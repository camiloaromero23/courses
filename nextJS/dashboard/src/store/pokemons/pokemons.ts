import { SimplePokemon } from "@/pokemons";
import { PayloadAction, createSlice } from "@reduxjs/toolkit";

export interface PokemonsState {
  favorites: {
    [key: string]: SimplePokemon;
  };
}

// const getInitialState = (): PokemonsState => {
//   // if (typeof localStorage === "undefined") return {}; // WARN: Messes with hydration

//   const pokemons = localStorage.getItem("pokemons");

//   return pokemons ? JSON.parse(pokemons) : {};
// };

const initialState: PokemonsState = {
  favorites: {}
  // ...getInitialState(),
  // "1": { id: "1", name: "bulbasaur" },
  // "3": { id: "3", name: "venusaur" },
  // "5": { id: "5", name: "charmeleon" },
};

const pokemonsSlice = createSlice({
  name: "pokemons",
  initialState,
  reducers: {
    setFavoritePokemons(state, action: PayloadAction<typeof initialState.favorites>) {
      state.favorites = action.payload;
    },

    toggleFavorite(state, action: PayloadAction<SimplePokemon>) {
      const pokemon = action.payload;

      if (!!state.favorites[pokemon.id]) {
        delete state.favorites[pokemon.id];
      } else {
        state.favorites[pokemon.id] = pokemon;
      }
    },
  },
});

export const { toggleFavorite, setFavoritePokemons } = pokemonsSlice.actions;
export default pokemonsSlice.reducer;
