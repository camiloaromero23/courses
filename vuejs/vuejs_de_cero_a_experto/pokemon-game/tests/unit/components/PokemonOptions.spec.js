import { shallowMount } from '@vue/test-utils';
import PokemonOptions from '@/components/PokemonOptions';

import { pokemons } from '../mocks/pokemons.mock';

describe('PokemonOptions components', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(PokemonOptions, {
      props: {
        pokemons,
        optionsEnabled: true,
      },
    });
  });

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should show the 4 options correctly', () => {
    const liTags = wrapper.findAll('li');

    expect(liTags.length).toBe(4);

    liTags.forEach((li, index) => {
      expect(li.text()).toBe(pokemons[index].name);
    });
  });

  it('should emit "selection" with its respective params once it is clicked', () => {
    const liTags = wrapper.findAll('li');

    liTags.forEach((li, index) => {
      li.trigger('click');
      expect(wrapper.emitted('pokemonSelected')[index]).toEqual([
        pokemons[index].id,
      ]);

      // console.log(wrapper.emitted('pokemonSelected').length);
    });
    expect(wrapper.emitted('pokemonSelected').length).toBe(4);
  });
});
