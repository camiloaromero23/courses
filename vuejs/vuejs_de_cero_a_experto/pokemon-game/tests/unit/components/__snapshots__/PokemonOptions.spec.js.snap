// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonOptions components should match the snapshot 1`] = `
<div class="options-container">
  <ul>
    <li disabled="true" class="">pikachu</li>
    <li disabled="true" class="">ivysaur</li>
    <li disabled="true" class="">venusaur</li>
    <li disabled="true" class="">charmander</li>
  </ul>
</div>
`;
